core/src/main/java/org/elasticsearch/script/ScriptService.java

```diff
@@ -150,323 +102,158 @@ public ScriptService(Settings settings, Environment env,
-        clusterSettings.addSettingsUpdateConsumer(SCRIPT_MAX_COMPILATIONS_PER_MINUTE, this::setMaxCompilationsPerMinute);
-    }
-
-    @Override
-    public void close() throws IOException {
-        IOUtils.close(scriptEngines);
-    }
-
-    private ScriptEngineService getScriptEngineServiceForLang(String lang) {
-        ScriptEngineService scriptEngineService = scriptEnginesByLang.get(lang);
-        if (scriptEngineService == null) {
-            throw new IllegalArgumentException("script_lang not supported [" + lang + "]");
-        }
-        return scriptEngineService;
-    }
-
-    private ScriptEngineService getScriptEngineServiceForFileExt(String fileExtension) {
-        ScriptEngineService scriptEngineService = scriptEnginesByExt.get(fileExtension);
-        if (scriptEngineService == null) {
-            throw new IllegalArgumentException("script file extension not supported [" + fileExtension + "]");
-        }
-        return scriptEngineService;
-    }
-
-    void setMaxCompilationsPerMinute(Integer newMaxPerMinute) {
-        this.totalCompilesPerMinute = newMaxPerMinute;
-        // Reset the counter to allow new compilations
-        this.scriptsPerMinCounter = totalCompilesPerMinute;
-        this.compilesAllowedPerNano = ((double) totalCompilesPerMinute) / TimeValue.timeValueMinutes(1).nanos();
-    }
-
-    /**
-     * Checks if a script can be executed and compiles it if needed, or returns the previously compiled and cached script.
-     */
-    public CompiledScript compile(Script script, ScriptContext scriptContext) {
-        Objects.requireNonNull(script);
-        Objects.requireNonNull(scriptContext);
-
-        ScriptType type = script.getType();
-        String lang = script.getLang();
-        String idOrCode = script.getIdOrCode();
-        Map<String, String> options = script.getOptions();
-
-        String id = idOrCode;
-
-        // lang may be null when looking up a stored script, so we must get the
-        // source to retrieve the lang before checking if the context is supported
-        if (type == ScriptType.STORED) {
-            // search template requests can possibly pass in the entire path instead
-            // of just an id for looking up a stored script, so we parse the path and
-            // check for appropriate errors
-            String[] path = id.split("/");
-
-            if (path.length == 3) {
-                if (lang != null && lang.equals(path[1]) == false) {
-                    throw new IllegalStateException("conflicting script languages, found [" + path[1] + "] but expected [" + lang + "]");
+        maxScriptSizeInBytes = ScriptService.SCRIPT_MAX_SIZE_IN_BYTES.get(settings);
+        this.scriptMetrics = scriptMetrics;
+        this.scriptPermits = new ScriptPermits(settings, scriptSettings, scriptContextRegistry);
+        this.compiler = new CachingCompiler<CacheKey>(settings, env, resourceWatcherService, scriptMetrics, "script") {
```

Comment:
Like an inner class for `ScriptService`? I think of this inner class as "the behavior in ScriptService that is distinct from TemplateService".